### DISCUSSION TITLE
listenSelf and StateProvider

### DISCUSSION THREAD

**Discussion by MarcinusX** [rrousselGit/riverpod#3873] [General]
`ref.listenSelf` got deprecated in favor of `Notifier.listenSelf` and I was using it in `StateProvider`.
```dart
final mainPageTab = StateProvider.autoDispose<MainPageTab>(
  (ref) {
    ref.listenSelf((_, next) => Analytics.setCurrentScreen(next.analyticsName));
    return MainPageTab.home;
  },
  name: 'mainPageTab',
);
```

Is there a way to use `listenSelf` for `StateProviders`?
I don't really want to create a `Notifier` that doesnt really add any extra value besides `StateProvider`.
If `Notifier` is the answer, does it mean that there is a plan to deprecate and remove `StateProvider` too?

**Comment by rrousselGit** [comment]
> If Notifier is the answer, does it mean that there is a plan to deprecate and remove StateProvider too?

Yes

There's no equivalent to StateProvider in codegen afterall ;)

**Comment by MarcinusX** [comment]
Got it, thanks :)

So I guess it's about time to let go old StateNotifiers and non-gen providers and move to recommended solutions 😅
Would you recommend upgrading straight from StateProviders, FutureProviders and StateNotifierProviders to code-gen or first migrating to self-written providers and then to code-gen?

**Reply by rrousselGit** [reply]
[This page](https://riverpod.dev/docs/concepts/about_code_generation) should summarise my though on it.

You're free to keep using Futureproviders

**Reply by MarcinusX** [reply]
Got it! Thanks! <3
